packages/core/src/tracing/clsHooked/async_local_storage_context.js

```diff
@@ -0,0 +1,324 @@
+   * returns a then-able, this behaves like runPromise. If not, this behaves like runAndReturn. In particular, no error
+   * is thrown if the given function does not return a promise.
+   *
+   * This function assumes that the returned promise (if any) is CLS-friendly or wrapped already.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runPromiseOrRunAndReturn(fn, ctx) {
+    let isPromise = false;
+    const context = ctx || this.createContext();
+    this.enter(context);
+
+    const valueOrPromise = fn(context);
+    isPromise = valueOrPromise && valueOrPromise.then && valueOrPromise.catch;
+    if (isPromise) {
+      return valueOrPromise;
+    }
+  }
+
+  /**
+   * Returns a wrapper around the given function which will enter CLS context which is active at the time of calling
+   * bind and leave that context once the function terminates. If no context is active, a new context will be created.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} context
+   */
+  bind(fn, context) {
+    context = context || storage.getStore() || this.createContext();
+
+    const self = this;
+
+    return function clsBind() {
+      self.enter(context);
+      return fn.apply(this, arguments);
+    };
+  }
+
+  /**
+   * Binds the given emitter to the currently active CLS context. Work triggered by an emit from that emitter will
+   * happen in that CLS context.
+   * @param {import('events').EventEmitter} emitter
+   */
+  bindEmitter(emitter) {
+    assert.ok(emitter.on && emitter.addListener && emitter.emit, 'can only bind real EEs');
+
+    const namespace = this;
+    const thisSymbol = `context@${this.name}`;
+
+    /**
+     * Capture the context active at the time the emitter is bound.
+     * @param {*} listener
+     */
+    function attach(listener) {
+      if (!listener) {
+        return;
+      }
+      if (!listener[CONTEXTS_SYMBOL]) {
+        listener[CONTEXTS_SYMBOL] = Object.create(null);
+      }
+
+      listener[CONTEXTS_SYMBOL][thisSymbol] = {
+        namespace,
+        context: storage.getStore()
+      };
+    }
+
+    /**
+     * At emit time, bind the listener within the correct context.
+     * @param {*} unwrapped
+     */
+    function bind(unwrapped) {
+      if (!(unwrapped && unwrapped[CONTEXTS_SYMBOL])) {
+        return unwrapped;
+      }
+
+      let wrapped = unwrapped;
+      const unwrappedContexts = unwrapped[CONTEXTS_SYMBOL];
+      Object.keys(unwrappedContexts).forEach(name => {
+        const thunk = unwrappedContexts[name];
+        wrapped = thunk.namespace.bind(wrapped, thunk.context);
+      });
+      return wrapped;
+    }
+
+    wrapEmitter(emitter, attach, bind);
+  }
+
+  /**
+   * @param {InstanaCLSContext} context
+   */
+  enter(context) {
+    assert.ok(context, 'context must be provided for entering');
+    storage.enterWith(context);
+    this.sharedContext = context;
```

Comment:
Wondering:

This class is a singleton. I do not fully understand how we can rely on `sharedContext` being set to the correct context.

1. GraphQL request mutation comes in. There are x subscribers
2. GraphQL pub/sub process is activated. We have to call `enter` in the GraphQL Subscription Instrumentation to somehow set the remembered/previous context.
3. We hope that no one else calls `enter` in between, otherwise we'd loose our shared context. As far as I can see there are other use cases where `enter` can be called in between (from other instrumentations), because GraphQL subscription instrumentation calling `enter` and the actual subscription tracing call in the GraphQL instrumentation is async!
4. GraphQL subscription is executed to inform the clients.
5. We hope `sharedContext` is still the correct one.

Reply:
Discussed: valid concern, but "this approach" is also used in the older context implementations.
We would like to change this, but need time & brain for that. I'll make that clear as part of a comment.